### PR TITLE
introduce blech32m format

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

**Comment by apoelstra** [comment]
Companion to https://github.com/ElementsProject/elements/pull/1025

File: src/address.rs [hunk 1]
```diff
@@ -769,4 +795,43 @@ mod test {
             roundtrips(&addr);
         }
     }
+
+    #[test]
+    fn test_blech32_vectors() {
+        // taken from Elements test/functional/rpc_invalid_address_message.py
```
Comment (LeoComandini) [inline src/address.rs:801]: Is there a reason why there is no `BLECH32M_VALID` test vector?

Comment (apoelstra) [reply src/address.rs:801]: There is also no `BECH32M_VALID` test vector. I simply converted all the existing Bitcoin test vectors (which exist to test the RPC error messages as much as they exist to test the decoding logic).

File: src/address.rs [hunk 2]
```diff
@@ -530,7 +556,7 @@ impl fmt::Display for Address {
                     data.extend_from_slice(&witprog);
                     let mut b32_data = vec![witver];
                     b32_data.extend_from_slice(&data.to_base32());
-                    blech32::encode_to_fmt(fmt, &hrp, &b32_data)
+                    blech32::encode_to_fmt(fmt, &hrp, &b32_data, blech32::Variant::Blech32)
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/address.rs]: Fixed

**Comment by apoelstra** [comment]
Updated for final bech32m version, fixed encoding bugs

File: src/blech32.rs [hunk 3]
```diff
@@ -144,23 +179,24 @@ pub fn decode(s: &str) -> Result<(&str, Vec<u5>), Error> {
...
 
     // Ensure checksum
-    if !verify_checksum(&hrp_bytes, &data) {
-        return Err(Error::InvalidChecksum);
-    }
+    match verify_checksum(&raw_hrp.as_bytes(), &data) {
+        Some(variant) => {
+            // Remove checksum from data payload
+            let dbl: usize = data.len();
+            data.truncate(dbl - 12);
```
Comment (stevenroose) [inline src/blech32.rs:186]: You lost the `// ELEMENTS: 6->12` comment here though. Is that still relevant or was it removed on purpose?

File: src/script.rs [hunk 4]
```diff
@@ -386,6 +386,16 @@ impl Script {
...
     }
 
+    /// Checks whether a script pubkey is a p2wsh output
+    #[inline]
+    pub fn is_v1plus_p2witprog(&self) -> bool {
+        self.0.len() > 1 &&
+        self.0.len() == self.0[1] as usize + 2 &&
+        self.0[0] >= opcodes::all::OP_PUSHNUM_1.into_u8() &&
+        self.0[0] <= opcodes::all::OP_PUSHNUM_16.into_u8() &&
+        self.0[1] <= opcodes::all::OP_PUSHBYTES_40.into_u8()
```
Comment (stevenroose) [inline src/script.rs:396]: It's implicit here that the PUSHBYTES start at 0. Probably fine but had to double check during review. Probably doesn't warrant adding an extra check. Perhaps a comment. But definitely a nit, not important.

**Review by stevenroose — COMMENTED** [review]
One ELEMENTS comment got lost, but not sure if it was removed on purpose.

File: src/blech32.rs (hunk 3 above)
Comment (apoelstra) [reply src/blech32.rs:186]: It is still relevant, I removed it by accident

**Comment by RCasatta** [comment]
on top of this I added a test to print a blech32m v1 address:

```
#[test]
fn blech32m_get_address_info() {
    let address = elements::Address {
        params: &AddressParams::LIQUID,
        payload: elements::address::Payload::WitnessProgram {
            version: u5::try_from_u8(1).unwrap(),
            program: vec![0u8; 32],
        },
        blinding_pubkey: Some(secp256k1_zkp::PublicKey::from_slice(&[2u8; 33]).unwrap()),
    };
    assert_eq!(address.to_string(),"lq1pqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7tgzuvqwaqsv");
}
```

Then on my liquid mainnet node I did: 

```
$ elements-cli getnetworkinfo | jq .subversion
"/Elements Core:0.21.0/"
$ elements-cli getaddressinfo lq1pqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7tgzuvqwaqsv
{
  "address": "lq1pqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7tgzuvqwaqsv",
  "scriptPubKey": "51200000000000000000000000000000000000000000000000000000000000000000",
  "ismine": false,
  "solvable": false,
  "iswatchonly": false,
  "iswitness": true,
  "witness_version": 1,
  "witness_program": "0000000000000000000000000000000000000000000000000000000000000000",
  "ischange": false,
  "labels": [
  ]
}
```

The field `confidential_key` is not printed

**Review by stevenroose — APPROVED** [review]
utACK

**Pull request merged**